### PR TITLE
xds: always use `::` and `IPv4Compact` for dynamic listener

### PR DESCRIPTION
**Pull request by zirain**
this's a huge changes seperated from https://github.com/envoyproxy/gateway/pull/4690.
we still need a PR to fix the bootstrap, especially for `admin` endpoint.

### CODE REVIEW

**Comment by zhaohuabing** [comment]
@zirain let's hold this after v1.2.2 because it changes too many files and make cherry-picking harder.

File: internal/gatewayapi/listener.go [hunk 1]
```diff
@@ -99,6 +100,10 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR resource
 			if !isReady {
 				continue
 			}
+
+			// EG always use `::` and set ipv4_compact with true to support both IPv4 and IPv6
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply internal/gatewayapi/listener.go:104]: Should we also use `::` and ipv4_compact for IPv4 only environment?

**Comment by arkodg** [comment]
> @zirain let's hold this after v1.2.2 because it changes too many files and would make cherry-picking harder.

@zhaohuabing imo this should of v1.2 since its part of the dual stack feature

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by zhaohuabing** [comment]
> > @zirain let's hold this after v1.2.2 because it changes too many files and would make cherry-picking harder.
> 
> @zhaohuabing imo this should of v1.2 since its part of the dual stack feature

OK, let's cherry-pick this to v1.2.2

File: internal/gatewayapi/listener.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply internal/gatewayapi/listener.go:104]: @zirain Could you please help me understand?
*  why prefer :: and ipv4_compact over addtional address
* would this approbach work with IPv4 only or underlying OS doesn't support IPv6?

[the remark this answers is not in the document]
Comment (zirain) [reply internal/gatewayapi/listener.go:104]: for my poor understanding, this's just like ` net.Listen("tcp", ":8000")` in golang, or maybe I'm wrong.

[the remark this answers is not in the document]
Comment (zhaohuabing) [reply internal/gatewayapi/listener.go:104]: > ipv4_compat
([bool](https://developers.google.com/protocol-buffers/docs/proto#scalar)) When binding to an IPv6 address above, this enables [IPv4 compatibility](https://tools.ietf.org/html/rfc3493#page-11). Binding to :: will allow both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into IPv6 space as ::FFFF:<IPv4-address>.

Accordinig to the Envoy docs, bindiing to :: will map IPv4 address into IPv6, so it probably won't work for IPv4 only.

**Review by zhaohuabing — APPROVED** [review]

**Comment by zhaohuabing** [comment]
LGTM thanks.

Non-blocking: Envoy now listens on all IP Families by default including IPv4 and IPv6, regardless of the IPFamily setting. It might be usefult to clarify this in the IPFamily docs to avoid potential confusion.

**Comment by zirain** [comment]
> LGTM thanks!
> 
> One non-blocking comment: Envoy now listens on all IP Families by default including IPv4 and IPv6, regardless of the IPFamily setting. It might be usefult to clarify this behavior in the IPFamily API docs to avoid potential confusion.

Will update these once we make an agreement on admin endpoint.

**Pull request merged**